File: Example/Lua/MySampleFilePlugin.lua
function on_file_processed(rel_file_path)
    -- 相对于原始读入文件夹的路径
    utils.logger:info(string.format("%s completed From Lua", rel_file_path.value))
end

function run()
    -- 这是 NormalJsonTranslator 要读取的原始 json 文件夹
    luaTranslator.m_inputDir = luaTranslator.m_projectDir / Path.new("gt_input")
    -- 这是 NormalJsonTranslator 翻译完成后的输出文件夹
    -- 重载了 Path 类型的 / 操作符，但要求其中至少一方是 Path 类型
    -- 用字符串对 Path 赋值可采用 varName = Path.new("str") 或 varName.value = "str"
    luaTranslator.m_outputDir = luaTranslator.m_projectDir / "gt_output"
    utils.logger:info(string.format("Current inputDir: %s", luaTranslator.m_inputDir))
    -- 有文件处理完毕后会回调这个函数
    luaTranslator.m_onFileProcessed = on_file_processed
    luaTranslator:normalJsonTranslator_run()

    -- -- Lua 没有多线程，所有函数都是阻塞的
    -- -- 但并不影响 TextPlugin，即使你的 conditionFunc 放在这个脚本里也没问题
    -- -- 因为 FilePlugin 和 TextPlugin 用的是不同的 LuaState(Lua环境)
    -- -- 如果想在 Lua 中使用 C++ NormalJsonTranslaor 的文件多线程
    -- -- 除过直接调用 run 函数之外
    -- -- 唯一的办法是使用 m_threadPool:push(luaTranslator , Vec{Paths}) (包括这个函数本身也是阻塞的)
    -- local relFilePaths = luaTranslator:normalJsonTranslator_beforeRun()
    -- if relFilePaths == nil then 
    --     utils.logger:info("可能是 DumpName 或 GenDict 之类无需 processFile 的 TransEngine")
    --     return
    -- end
    -- local relFilePathsStr = {}
    -- for i = 1, #relFilePaths do
    --     table.insert(relFilePathsStr, relFilePaths[i].value)
    -- end
    -- utils.logger:info("relFilePaths: \n" .. table.concat(relFilePathsStr, "\n"))
    -- luaTranslator.m_threadPool:resize(35)
    -- luaTranslator.m_controller:makeBar(luaTranslator.m_totalSentences, 35)
    -- luaTranslator.m_threadPool:push(luaTranslator, relFilePaths)
    -- luaTranslator:normalJsonTranslator_afterRun()

    -- luaTranslator:epubTranslator_beforeRun()
    -- -- std::function<void(fs::path)>
    -- orgOnFileProcessedInEpubTranslator = luaTranslator.m_onFileProcessed
    -- luaTranslator.m_onFileProcessed = function (relFilePathProcessed)
    --     utils.logger:info("拦截然后加一条日志喵")
    --     orgOnFileProcessedInEpubTranslator(relFilePathProcessed)
    -- end
    -- luaTranslator:normalJsonTranslator_run()
    
end

function init()
    utils.logger:info("MySampleFilePluginFromLua starts")
    utils.logger:info("apiStrategy: " .. luaTranslator.m_apiStrategy)
    local tomlConfig = toml.parse(luaTranslator.m_projectDir / "config.toml")
    if tomlConfig == nil then
        utils.logger:info("出错错了喵")
    else
        local epubPreReg1 = tomlConfig.plugins.Epub.preprocRegex[1]
        utils.logger:info("{epubPreReg1} org: " .. epubPreReg1.org .. ", rep: " .. epubPreReg1.rep)
    end
end

function countMap(map)
    local count = 0
    for k, v in pairs(map) do
        count = count + 1
    end
    return count
end

function unload()
    utils.logger:info("MySampleFilePluginFromLua unloads")
    -- std::map<fs::path, std::map<fs::path, bool>> m_jsonToSplitFileParts;
    local partsTable = luaTranslator.m_jsonToSplitFileParts
    local strs = {}
    for jsonPath, filePartsMap in pairs(partsTable) do
        for splitFilePart, comp in pairs(filePartsMap) do
            if comp then
                table.insert(strs, splitFilePart.value)
                local j = json.parse(luaTranslator.m_cacheDir / splitFilePart)
                if j == nil then
                    utils.logger:info("出错错了喵")
                elseif #j >= 1 then
                    table.insert(strs, j[1].translated_preview)
                end
            else
                utils.logger:error("不应该发生")
            end
        end
    end
    utils.logger:info("map keys: \n" .. table.concat(strs, "\n"))

    -- -- 为了能使用 pairs 遍历，map返回类型都是副本，如果想修改，必须 luaTranslator.mapVar = copy
    -- utils.logger:info("原有 " .. tostring(countMap(luaTranslator.m_jsonToSplitFileParts)) .. " 个 key-value 对")
    -- local newPartTable = { MyNewPartPath1 = false, ["MyNewPartPath2"] = false }
    -- luaTranslator.m_jsonToSplitFileParts[Path.new("MyNewJsonPathKey")] = newPartTable
    -- utils.logger:info("依然为 " .. tostring(countMap(luaTranslator.m_jsonToSplitFileParts)) .. " 个 key-value 对")
    -- partsTable[Path.new("MyNewJsonPathKey")] = newPartTable
    -- luaTranslator.m_jsonToSplitFileParts = partsTable
    -- utils.logger:info("现有 " .. tostring(countMap(luaTranslator.m_jsonToSplitFileParts)) .. " 个 key-value 对")

    -- local jsonToInfoMap = luaTranslator.m_jsonToInfoMap
    -- for jsonPath, jsonInfo in pairs(jsonToInfoMap) do
    --     utils.logger:info("metadataVecSize: " .. tostring(#jsonInfo.metadata))
    --     for i=1, #jsonInfo.metadata do
    --         utils.logger:info("offset: " .. tostring(jsonInfo.metadata[i].offset) .. ", length: " .. tostring(jsonInfo.metadata[i].length))
    --     end
    -- end

    utils.logger:info("MySampleFilePluginFromLua unloads")
end

File: Example/Python/MySampleFilePlugin.py
# Required: Import C++ binding module
import gpp_plugin_api as gpp
from pathlib import Path
import tomllib
import threading
import queue
import time

# See GalTranslPP/PythonManager.cpp for all registered types and functions

# Required: Declare pythonTranslator
# C++ will assign base class pointer to this attribute before init
pythonTranslator = None

logger = None

def process_file(rel_file_path, worker_id):
    """Wrapper function, worker_id is the thread number"""
    try:
        pythonTranslator.m_controller.addThreadNum()
        pythonTranslator.processFile(rel_file_path, worker_id)
        pythonTranslator.m_controller.reduceThreadNum()
    except Exception as e:
        error_msg = f"[Thread {worker_id}] Error processing file {rel_file_path}: {str(e)}\n{traceback.format_exc()}"
        logger.error(error_msg)

def worker(worker_id, task_queue):
    """Worker thread function, worker_id is the fixed ID for this thread"""
    while True:
        task = task_queue.get()
        if task is None:  # Received exit signal
            task_queue.task_done()
            break

        try:
            rel_file_path = task  # Task only contains file path
            process_file(rel_file_path, worker_id)
        finally:
            task_queue.task_done()

def multi_threads_run(relFilePaths):
    # Multi-threading uses manual thread management
    # Because using multiprocessing or ThreadPoolExecutor in sub-interpreters has some quirky issues
    # Without LLM translation, behavior approaches single-threaded because I only release GIL on C++ side during LLM translation
    MAX_WORKERS = 29
    pythonTranslator.m_controller.makeBar(pythonTranslator.m_totalSentences, MAX_WORKERS)
    task_queue = queue.Queue()
    threads = []

    # Create and start worker threads, each with a fixed worker_id
    for worker_id in range(MAX_WORKERS):
        t = threading.Thread(
            target=worker,
            args=(worker_id, task_queue),  # Pass worker_id
            name=f"Worker-{worker_id}"
        )
        t.daemon = False
        t.start()
        threads.append(t)
        logger.info(f"Started worker thread {worker_id}")

    # Put all tasks into queue
    for relFilePath in relFilePaths:
        task_queue.put(relFilePath)
    logger.info(f"Added {len(relFilePaths)} tasks to queue")

    # Wait for all tasks to complete
    task_queue.join()

    # Send exit signal to all worker threads
    for _ in range(MAX_WORKERS):
        task_queue.put(None)

    # Wait for all threads to finish
    for t in threads:
        t.join(timeout=30)
        if t.is_alive():
            logger.warn(f"Thread {t.name} did not terminate properly")

    logger.info("All Python threads finished execution")

def run():
    # Each translator's filePlugin and textPlugin uses a dedicated sub-interpreter (except NLP tokenize series)
    # So this function can block without affecting other translator's work
    relFilePaths = pythonTranslator.normalJsonTranslator_beforeRun()
    if relFilePaths is None:
        logger.info("Probably a TransEngine like DumpName or GenDict that doesn't need processFile")
        return
    # Simple single-threaded processing
    # for relFilePath in relFilePaths:
    #     pythonTranslator.processFile(relFilePath, 0)
    multi_threads_run(relFilePaths)
    pythonTranslator.normalJsonTranslator_afterRun()
    return

    # Small example inheriting from Epub, also recommended to read Lua section first
    pythonTranslator.epubTranslator_beforeRun()
    # std::function<void(fs::path)>
    orgOnFileProcessedInEpubTranslator = pythonTranslator.m_onFileProcessed
    # Like Lua, this can also be a closure
    # But if you set this, then you must NOT use xxTranslator_process() series functions that create threads on C++ side
    # Otherwise it will cause deadlock because C++ side threads don't have GIL and can't call Python closures
    def new_call_back(rel_file_path_processed):
        logger.info("Intercepting to add a log entry meow")
        orgOnFileProcessedInEpubTranslator(rel_file_path_processed)
    pythonTranslator.m_onFileProcessed = new_call_back
    relFilePaths = pythonTranslator.normalJsonTranslator_beforeRun()
    if relFilePaths is None:
        logger.info("Probably a TransEngine like DumpName or GenDict that doesn't need processFile")
        return
    multi_threads_run(relFilePaths)
    pythonTranslator.normalJsonTranslator_afterRun()
    return

def init():
    # TOML config files and dictionary files are loaded after init
    pythonTranslator.normalJsonTranslator_init()
    # Recommended to read Lua plugin documentation first
    # pythonTranslator.epubTranslator_init()

    logger.info("MySampleFilePluginFromPython starts")
    logger.info("Current inputDir: " + str(pythonTranslator.m_inputDir))
    tomlPath = pythonTranslator.m_projectDir / "config.toml"

    with open(tomlPath, "rb") as f:
        tomlConfig = tomllib.load(f)
        epubPreReg1 = tomlConfig['plugins']['Epub']['preprocRegex'][0]
        logger.info("{epubPreReg1} org: " + epubPreReg1['org'] + ", rep: " + epubPreReg1['rep'])

def unload():
    logger.info("MySampleFilePluginFromPython unloads")
    # std::map<fs::path, std::map<fs::path, bool>> m_jsonToSplitFileParts;
    partsTable = pythonTranslator.m_jsonToSplitFileParts
    strs = []
    for jsonPath, filePartsMap in partsTable.items():
        for splitFilePart, completed in filePartsMap.items():
            if completed:
                strs.append(str(splitFilePart))
            else:
                logger.error("This should not happen")
    result = "\n".join(strs)
    # logger.info("map keys: \n" + result)
